openedx-unsupported/configuration
# Converts the seed job xml to a groovy DSL

Replaces templates/seed_job_template.xml with templates/seedJob.groovy, which will be easier to maintain.

However, see notes below for the drawbacks.

**JIRA tickets**: Implements [OLIVE-34](https://openedx.atlassian.net/browse/OLIVE-34), [OC-1528](https://tasks.opencraft.com/browse/OC-1528)

**Dependencies**: None

**Testing instructions**:

_Note: If you've already created a Jenkins Analytics vagrant instance to test PR #2959 , then you can just checkout this branch and run `vagrant provision`._

To test via vagrant:
1. Install edx/configuration requirements in a virtualenv:
   
   ```
   virtualenv venv
   . venv/bin/activate
   git clone https://git@github.com/open-craft/configuration.git
   cd configuration
   git checkout -b jill/olive-34-seed-job-groovy origin/jill/olive-34-seed-job-groovy
   make requirements
   ```
2. Create a YAML file containing the jenkins analytics configuration.
   - See [jenkins_analytics README](https://github.com/open-craft/configuration/blob/jill/olive-34-seed-job-groovy/playbooks/roles/jenkins_analytics/README.md) for details.
   - Also see [example ansible-jenkins-credentials.yml.zip](https://github.com/edx/configuration/files/219328/seed.job.ansible-jenkins-credentials.yml.zip)
3. Verify that you have access to https://github.com/edx-ops/edx-jenkins-job-dsl/
4. Build the vagrant instance, using the YAML file created above:
   
   ```
   cd vagrant/base/analytics_jenkins
   export VAGRANT_JENKINS_LOCAL_VARS_FILE=/path/to/your.yml
   vagrant up
   ```
5. Should run successfully, building the AnalyticsSeedJob successfully as the last step.

**Author notes and concerns**:

I had issues running the seedJob.groovy file, and I'm not happy with the solution I ended up with.  Would appreciate any ideas, if you have time.
- Tried running directly via Jenkins CLI groovy and the [jenkins /script console](https://wiki.jenkins-ci.org/display/JENKINS/Jenkins+Script+Console).  I was able to import javaposse.jobdsl classes, but unable to create any jobs?  Got errors like `groovy.lang.MissingMethodException: No signature of method: RemoteClass.FreeStyleJob()`.  
- Tried wrapping it in a call to javaposse.jobdsl.Run.main and running via Jenkins
  CLI.  But this class tries to write to './AnalyticsSeedJob.xml', and errors out, even when running as root!  Baffled.

So instead, tasks/main.yml clones the job-dsl-plugin repo, and generates the
job XML using the instructions in the [job-dsl wiki](https://github.com/jenkinsci/job-dsl-plugin/wiki/User-Power-Moves).  The  build takes 3-7 min the first time, and <30 sec each subsequent run.

**Reviewers**
- [ ] @jibsheet
